File: Data Structures & Algorithms/middle-of-the-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # problem is that we do not know the total length of the LL
        # we can create a ptr that points to the end of the LL
        temp = head
        curr = head
        count = 1

        while temp:
            if temp.next == None:
                break

            temp = temp.next
            count+=1
        
        mid = count // 2
        print(mid)

        for i in range(mid):
            curr = curr.next
        
        return curr
            

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* ctr = head;
        int count = 0;

        while (ctr) {
            count++;
            ctr = ctr -> next;
        }
        
        ListNode* temp = head;
        ListNode* prev = nullptr;
        int remove_pos = count - n;
        for (int i = 0; i < count; ++i) {
            if (i == remove_pos) {
                if (prev == nullptr) {
                    ListNode* newHead = temp -> next;
                    return newHead;
                } else {
                    prev -> next = temp -> next;
                    return head;
                }
            }
            prev = temp;
            temp = temp -> next;
        }

        return head;
    }
};
